### PR TITLE
Add the audit pyCADF middleware for audit logging

### PR DESCRIPTION
**Pull request by edtubillara**
The pyCADF middleware will be enabled by default to have audit events logged
in the API logs for the following services:
    ceilometer, cinder, glance, heat, neutron, and nova

### CODE REVIEW

File: roles/cinder-common/templates/etc/cinder/cinder_api_audit_map.conf [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+os-quota-specs = project
+qos-specs = qos-spec
+snapshots = snapshot
+types = type
+volumes = volume
+
+# map endpoint type defined in service catalog to CADF typeURI
+[service_endpoints]
+volume = service/storage/block
+volumev2 = service/storage/block
```
Comment (j2sol) [inline roles/cinder-common/templates/etc/cinder/cinder_api_audit_map.conf]: Where are these contents coming from? Seems like a thing that could drift with each major openstack release.

**Comment by j2sol** [comment]
ok to test

**Comment by j2sol** [comment]
Do we know how much this will inflate the logs?

**Comment by edtubillara** [comment]
So it should inflate the logs quite a bit, whenever there is an action for a user (including a read), 
it will put an audit event entry in the api log (...nova-api.log). Several audit read events can happen when a user visits a section of a page in horizon (like the instances page).

The format per event (and the size) is similar to what happens for the CADF notifications for keystone:
https://developer.ibm.com/opentech/2015/07/06/viewing-keystone-cadf-notifications-with-ceilometer-and-rabbitmq/

So there may be some maintenance for the correctly mapping the user requests to correct audit actions. That is what the mapping file is used for. The API latest mapping configuration files can be found here:
https://github.com/openstack/pycadf/tree/master/etc/pycadf

Some more documentation here:
http://docs.openstack.org/developer/keystonemiddleware/audit.html

By the way I need to update the pull request, it seems that heat breaks.

**Comment by j2sol** [comment]
This needs to be feature flagged, and disabled by default. That'll allow us to turn it on slowly and observe what kind of impact it has.

**Comment by edtubillara** [comment]
ok sounds good. I will update and make it a feature flag.

**Comment by edtubillara** [comment]
I have changed it so that it should be feature flagged. By default it is set to False. 
To enable for a service (such as cinder), the user will need to set
the following values in all.yml-
cinder:
  auditing:
    enabled: True
    logging: True

A third value for messaging may be placed in a future request for putting the audit event on the message queue for rabbit/ceilometer to consume.

**Comment by edtubillara** [comment]
I looked at the Jenkins build and it looks like it failed due to fetch failures. Can someone run it again?

**Comment by j2sol** [comment]
retest

**Comment by j2sol** [comment]
Can you change the if statements to use the `|bool` filter? That will ensure that the value set for those variables is filtered through to ensure it's a boolean.  See https://github.com/blueboxgroup/ursula/blob/master/roles/keystone/templates/etc/keystone/keystone-paste.ini#L66 as an example.

**Comment by edtubillara** [comment]
Thanks for the review! I have added the bool filter.

**Pull request merged**